src/pybind/mgr/orchestrator_cli/module.py

```diff
@@ -23,48 +42,126 @@ def inner(*args, **kwargs):
             return HandleCommandResult(-errno.ENOENT, stderr=str(e))
     return inner
 
+# Reasons for storing generators in stead of completions are:
+# 1. some CLI commands do some (very valid) massaging with the result
+# 2. some CLI commands may use more than one completion
+# 3. Using an alternative appoach would split up the synchronous code into asynchronous completion
+#    handlers. And I don't like asynchronous completion handlers
+#
+# The  alternative would be to use something like _Completion.next() that executes the
+# massaging.
+
+
+GenComp = namedtuple('GenComp', ['name', 'gen', 'comp'])
+
+
+def _handle_generator(self, should_wait, name, it):
+    # type: (orchestrator.OrchestratorClientMixin, bool, Generator[Union[orchestrator._Completion, HandleCommandResult]]) -> HandleCommandResult
+    c = next(it)
+    if should_wait:
+        while not isinstance(c, HandleCommandResult):
+            self._orchestrator_wait([c])
+            c = it.send(c)
+
+        it.close()
+        return c
+    else:
+        self._completions.append(GenComp(name, it, c))
+        return HandleCommandResult(stderr='queued.')
+
+
+def handle_generator(func):
+    @wraps(func)
+    def inner(self, *args, **kwargs):
+        wait = kwargs.pop('wait', True)
+        return _handle_generator(self, wait, func.__name__, func(self, *args, **kwargs))
+    return inner
+
 
 class OrchestratorCli(orchestrator.OrchestratorClientMixin, MgrModule):
     MODULE_OPTIONS = [
         {'name': 'orchestrator'}
     ]
 
+    def __init__(self, *args, **kwargs):
+        super(OrchestratorCli, self).__init__(*args, **kwargs)
+        self._completions = []  # type: List[GenComp]
+
     def _select_orchestrator(self):
         return self.get_module_option("orchestrator")
 
+    @CLIWriteCommand('orchestrator wait',
+                     "name=refs,type=CephString,req=false,n=N",
+                     'Waits for all running completions')
+    @handle_exceptions
+    def _wait(self, refs=None):
```

Comment:
This waits synchronously in the event handler thread without a possibility to time-out. Relates to https://tracker.ceph.com/issues/37514